saved-sites/saved-sites-impl/src/main/java/com/duckduckgo/savedsites/impl/sync/algorithm/SavedSitesSyncPersisterAlgorithm.kt

```diff
@@ -80,27 +81,39 @@ class RealSavedSitesSyncPersisterAlgorithm @Inject constructor(
         // check all children, the ones that are not in allFolders don't have a parent
         val foldersWithoutParent = allFolderIds.filterNot { allChildren.contains(it) }
         foldersWithoutParent.forEach { folderId ->
-            if (savedSitesRepository.getFolder(folderId) != null) {
+            val folder = savedSitesRepository.getFolder(folderId)
+            if (folder != null) {
                 processIds.add(folderId)
             }
-            processFolder(folderId, SavedSitesNames.BOOKMARKS_ROOT, bookmarks.entries, lastModified, processIds, conflictResolution)
+            if (folder != null && folder.modifiedAfter(clientModifiedSince) && conflictResolution != DEDUPLICATION) {
```

Comment:
these functions are recursive, I can’t check for this condition inside because otherwise the values won’t be correct.
we check for deletion afterwards, not here. 
we can probably make the algorithm more readable, but it’s not something I’d like to do now.

Reply:
let's forget readability then.
At least I would say we need to be consistent with the pixel in other models, and if delete shouldn't trigger the pixel, we should not count it here, is that a correct assumption?

Reply:
`processFolder` will never process a deleted item. A deleted item won’t trigger a pixel. What made you think it does? What is not consistent?

Reply:
> In line 65 -> val allFolders = bookmarks.entries.filter { it.isFolder() } a deleted item won’t be returned here. processFolder will never process a deleted item.

I don't see how that is true. From what I see we process folders and bookmarks, including deleted ones, which is correct. Otherwise we wouldn't delete them locally if we don't process them.

Reply:
A deleted entity looks like this:
```
      {
        "deleted": "2023-05-23T11:35:01Z",
        "id": "2E085BF4-EB03-4B71-8848-5C1A30DCE53E",
        "last_modified": null
      },
```

If you look at the start of the function, you’ll see this:

```
val allFolders = bookmarks.entries.filter { it.isFolder() }
val allBookmarkIds = bookmarks.entries.filter { it.isBookmark() }.map { it.id }
```

Both of these won’t return deleted entities, because `isFolder` and `isBookmark` check for the `folder` and `page` field that are not present in deleted entities. For that reason, `processFolder` and `processBookmark` won’t iterate over deleted entities.

After we’ve processed non deleted items, we have this:

```
val allDeletedIds = bookmarks.entries.filter { it.deleted != null }.map { it.id }
processDeletedItems(allDeletedIds)

```
It’s here where we iterate over the deleted entities, and delete them. Why? Because it’s a tree structure, we can’t delete from the top. Subfolders might be deleted, but it’s content might have moved, we only know that after processing all the items in the payload. This is different in other models, where there’s no relation between entities. 

Going back to the original question.  We process deleted entities, and we don’t trigger the pixel for those entities.

Reply:
gotcha! thanks for explaining ⭐️ 

I was under the assumption `isFolder` or `isBookmark` were not filtering deteled items, now it's clear.